pkg/cni/plugin.go

```diff
@@ -0,0 +1,259 @@
+
+		// TODO: duplicate following lines for ipv6 support, when it will be added in other places
+		ipv4SysctlValueName := fmt.Sprintf(IPv4InterfaceArpProxySysctlTemplate, ifaceNameAllowingDots)
+		if _, err := sysctl.Sysctl(ipv4SysctlValueName, "1"); err != nil {
+			// remove the newly added link and ignore errors, because we already are in a failed state
+			netlink.LinkDel(macvtapLink)
+			return fmt.Errorf("failed to set proxy_arp on newly added interface %q: %v", macvtapLink.Attrs().Name, err)
+		}
+		return nil
+	})
+	return err
+}
+
+func configureMacvtap(conf *NetConf, macAddr net.HardwareAddr, ifName string, netns ns.NetNS) (*current.Interface, error) {
+	macvtapIface, err := netlink.LinkByName(conf.DeviceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup device %q: %v", conf.DeviceID, err)
+	}
+
+	// move the macvtap interface to the pod's netns
+	if err := netlink.LinkSetNsFd(macvtapIface, int(netns.Fd())); err != nil {
+		return nil, fmt.Errorf("failed to move iface %s to the netns %d because: %v", macvtapIface, netns.Fd(), err)
+	}
+
+	// configure the MTU of the macvtap iface
+	err = netns.Do(func(_ ns.NetNS) error {
+		if conf.MTU != 0 {
+			if err := netlink.LinkSetMTU(macvtapIface, conf.MTU); err != nil {
+				return fmt.Errorf("failed to set the macvtap MTU for %s: %v", conf.DeviceID, err)
+			}
+		}
+
+		if macAddr.String() != "" {
+			if err = netlink.LinkSetHardwareAddr(macvtapIface, macAddr); err != nil {
+				return fmt.Errorf("failed to add hardware addr to %q: %v", ifName, err)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	macvtap := &current.Interface{Name: ifName}
+	err = configureArp(macvtapIface, netns)
+	if err != nil {
+		return nil, err
+	}
+
+	err = renameMacvtapIface(macvtapIface, macvtap, ifName, netns)
+	if err != nil {
+		return nil, err
+	}
+
+	return macvtap, err
+}
+
+func renameMacvtapIface(macvtapLink netlink.Link, macvtapIface *current.Interface, ifaceName string, netns ns.NetNS) error {
+	err := netns.Do(func(_ ns.NetNS) error {
+		err := ip.RenameLink(macvtapLink.Attrs().Name, ifaceName)
+		if err != nil {
+			_ = netlink.LinkDel(macvtapLink)
+			return fmt.Errorf("failed to rename macvlan to %q: %v", ifaceName, err)
+		}
+
+		updatedLink := macvtapLink
+		updatedLink.Attrs().Name = ifaceName
+
+		// Re-fetch macvlan to get all properties/attributes
+		contMacvlan, err := netlink.LinkByName(ifaceName)
+		if err != nil {
+			return fmt.Errorf("failed to refetch macvlan %q: %v", ifaceName, err)
+		}
+		macvtapIface.Mac = contMacvlan.Attrs().HardwareAddr.String()
+		macvtapIface.Sandbox = netns.Path()
+
+		return nil
+	})
+	return err
+}
+
+// CmdAdd - CNI interface
+func CmdAdd(args *skel.CmdArgs) error {
+	var err error
+	netConf, cniVersion, err := loadConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	envArgs, err := getEnvArgs(args.Args)
+	if err != nil {
+		return err
+	}
+
+	netns, err := ns.GetNS(args.Netns)
```

Comment:
This can go after parsing of MAC. Let's work with the namespace for as little time as we can.

Reply:
Done.